DDR command shrc broken

DDR needs to be backward compatible so that the most recent version of DDR can be used to review old system core files.

The system core file was generated from pmz6480sr5fp27-20190104_01(SR5 FP27).  The DDR I am using was a recent one (build 428505, built on September 30,2019).  This is the error I got:

> shrc incache 0x000002005FB89028
!j9sharedclassconfig 0x00000050086D6B60

Error executing DDR command !shrc incache 0x000002005FB89028 : null
java.lang.reflect.InvocationTargetException
        at sun.reflect.GeneratedMethodAccessor5.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:55)
        at java.lang.reflect.Method.invoke(Method.java:508)
        at com.ibm.jvm.dtfjview.CombinedContext.executeDDRInteractiveCommand(CombinedContext.java:265)
        at com.ibm.jvm.dtfjview.CombinedContext.execute(CombinedContext.java:175)
        at com.ibm.jvm.dtfjview.Session.execute(Session.java:830)
        at com.ibm.jvm.dtfjview.Session.execute(Session.java:785)
        at com.ibm.jvm.dtfjview.DTFJView.process(DTFJView.java:68)
        at com.ibm.jtc.service.tools.jdmpview.JdmpviewTool.process(JdmpviewTool.java:81)
        at com.ibm.jtc.service.store.Store.process(Store.java:131)
        at com.ibm.jtc.service.store.Store.processSimpleCommand(Store.java:86)
        at com.ibm.jtc.service.store.input.Tree.processObject(Tree.java:278)
        at com.ibm.jtc.service.store.input.Tree.process(Tree.java:217)
        at com.ibm.jtc.service.store.Store.runInteractive(Store.java:409)
        at com.ibm.jtc.service.store.Store.open(Store.java:308)
        at com.ibm.jtc.service.store.Store.main(Store.java:56)
Caused by: java.lang.NoSuchFieldError: com/ibm/j9ddr/vm29/structure/J9SharedClassCacheDescriptor._previousOffset_
        at com.ibm.j9ddr.vm29.pointer.generated.J9SharedClassCacheDescriptorPointer.previous(J9SharedClassCacheDescriptorPointer.java:243)
        at com.ibm.j9ddr.vm29.tools.ddrinteractive.commands.ShrCCommand.dbgShrcReadConfig(ShrCCommand.java:1160)
        at com.ibm.j9ddr.vm29.tools.ddrinteractive.commands.ShrCCommand.getSharedCacheMetadataStart(ShrCCommand.java:1282)
        at com.ibm.j9ddr.vm29.tools.ddrinteractive.commands.ShrCCommand.run(ShrCCommand.java:103)
        at com.ibm.j9ddr.tools.ddrinteractive.Context.tryCommand(Context.java:229)
        at com.ibm.j9ddr.tools.ddrinteractive.Context.execute(Context.java:202)
        at com.ibm.j9ddr.tools.ddrinteractive.DDRInteractive.execute(DDRInteractive.java:356)
        at com.ibm.j9ddr.command.CommandReader.processLine(CommandReader.java:79)
        at com.ibm.j9ddr.tools.ddrinteractive.DDRInteractive.processLine(DDRInteractive.java:331)
        ... 16 more

--- a/debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/ShrCCommand.java
+++ b/debugtools/DDR_VM/src/com/ibm/j9ddr/vm29/tools/ddrinteractive/commands/ShrCCommand.java
@@ -78,6 +78,7 @@ public class ShrCCommand extends Command
 	private static final String rangeDelim = "..";
 	private static long cacheTotalSize = 0;
 	private static final long TYPE_PREREQ_CACHE = J9ConstantHelper.getLong(ShcdatatypesConstants.class, "TYPE_PREREQ_CACHE", -1);
+	private static final long J9SHR_DATA_TYPE_STARTUP_HINTS = J9ConstantHelper.getLong(ShCFlags.class, "J9SHR_DATA_TYPE_STARTUP_HINTS", -1);
 
 
 	public ShrCCommand()
@@ -1127,7 +1128,11 @@ public class ShrCCommand extends Command
 			CommandUtils.dbgPrint(out, "CHARARRAY data %d metadata %d\n", chararrayDataLen, chararrayMetaLen);
 			CommandUtils.dbgPrint(out, "BYTEDATA Summary\n");
 			CommandUtils.dbgPrint(out, "\tUNKNOWN %d  HELPER %d  POOL %d  AOTHEADER %d\n", numByteOfType[(int) J9SHR_DATA_TYPE_UNKNOWN], numByteOfType[(int) J9SHR_DATA_TYPE_HELPER], numByteOfType[(int) J9SHR_DATA_TYPE_POOL], numByteOfType[(int) J9SHR_DATA_TYPE_AOTHEADER]);
-			CommandUtils.dbgPrint(out, "\tJCL %d  VM %d  ROMSTRING %d  ZIPCACHE %d  STARTUPHINTS %d\n", numByteOfType[(int) J9SHR_DATA_TYPE_JCL], numByteOfType[(int) J9SHR_DATA_TYPE_VM], numByteOfType[(int) J9SHR_DATA_TYPE_ROMSTRING], numByteOfType[(int) J9SHR_DATA_TYPE_ZIPCACHE], numByteOfType[(int) J9SHR_DATA_TYPE_STARTUP_HINTS]);
+			if (J9SHR_DATA_TYPE_STARTUP_HINTS < 0) {
+				CommandUtils.dbgPrint(out, "\tJCL %d  VM %d  ROMSTRING %d  ZIPCACHE %d\n", numByteOfType[(int) J9SHR_DATA_TYPE_JCL], numByteOfType[(int) J9SHR_DATA_TYPE_VM], numByteOfType[(int) J9SHR_DATA_TYPE_ROMSTRING], numByteOfType[(int) J9SHR_DATA_TYPE_ZIPCACHE]);
+			} else {
+				CommandUtils.dbgPrint(out, "\tJCL %d  VM %d  ROMSTRING %d  ZIPCACHE %d  STARTUPHINTS %d\n", numByteOfType[(int) J9SHR_DATA_TYPE_JCL], numByteOfType[(int) J9SHR_DATA_TYPE_VM], numByteOfType[(int) J9SHR_DATA_TYPE_ROMSTRING], numByteOfType[(int) J9SHR_DATA_TYPE_ZIPCACHE], numByteOfType[(int) J9SHR_DATA_TYPE_STARTUP_HINTS]);
+			}
 			CommandUtils.dbgPrint(out, "\tJITHINT %d  AOTCLASSCHAIN %d AOTTHUNK %d\n", numByteOfType[(int) J9SHR_DATA_TYPE_JITHINT], numByteOfType[(int) J9SHR_DATA_TYPE_AOTCLASSCHAIN], numByteOfType[(int) J9SHR_DATA_TYPE_AOTTHUNK]);
 			if (cacheletMetaLen > 0) {
 				CommandUtils.dbgPrint(out, "CACHELET count %d (without segments %d) metadata %d\n", numCachelets, numCacheletsNoSegments, cacheletMetaLen);
@@ -1156,8 +1161,11 @@ public class ShrCCommand extends Command
 
 		SH_CacheMapPointer cacheMap = sharedClassConfig.sharedClassCache();
 		J9SharedClassCacheDescriptorPointer cacheDescriptor = sharedClassConfig.cacheDescriptorList();
-		/* get cacheDescriptor of layer 0 cache */
-		cacheDescriptor = cacheDescriptor.previous();
+		
+		if (topLayer > 0) {
+			/* get cacheDescriptor of layer 0 cache */
+			cacheDescriptor = cacheDescriptor.previous();
+		}
 		int layer = 0;
 		do {
 			/* TRY TO GET : cacheStartAddress */
@@ -1241,7 +1249,9 @@ public class ShrCCommand extends Command
 			if (0 == segmentPtr[layer].longValue()) {
 				CommandUtils.dbgPrint(out, "segmentPtr is zero for layer %d\n", layer);
 			}
-			cacheDescriptor = cacheDescriptor.previous();
+			if (topLayer > 0) {
+				cacheDescriptor = cacheDescriptor.previous();
+			}
 			layer += 1;
 		} while (layer <= topLayer);
 		
@@ -1909,7 +1919,7 @@ public class ShrCCommand extends Command
 			int topLayer = dbgShrcCacheTopLayer(out, sharedClassConfig);
 			int layer = (-1 == topLayer) ? -1 : 0;
 			do {
-				found = dbgShrcHeaderOperations(out, cacheStartAddressArray[layer], address, -1, layer).getV1();
+				found = dbgShrcHeaderOperations(out, (layer >= 0) ? cacheStartAddressArray[layer] : cacheStartAddress, address, -1, layer).getV1();
 				if (found) {
 					break;
 				} else if (-1 == layer) {

